Fix exitcode 127 error

kashizui opened the pull request:
Update docker container preamble commands to use chown instead of sudo to ensure proper file permissions. Most newer docker images don't have the sudo command, causing a 'command not found' error.

Also added an additional docker request to fetch container logs at the end of a run to check for any error messages that were previously missed (such as 'sudo: command not found').

Fixes codalab/codalab-worksheets#224

kashizui committed:
Fix empty-relationship check (#624)

Fixes #619

kashizui committed:
Increase request body size limit to 8 MiB (#631)

Fixes #627

kashizui committed:
Support nvidia-docker (#626)

Use nvidia-docker-plugin if available to set up the container creation request for GPU support. Otherwise, fall back to original hacks for GPU support.

Fixes #591

kashizui committed:
Remove unnecessary client dependencies (#636)

Fixes #521

maxwang7 committed:
Fix for displaying [Object object] for empty files (#628)

Checks for empty list (content). If empty, sets info to empty string.

yashsavani committed:
Send notifications (#604)

* working on back end database.

* working send_notifications on signup.

* changed tests to work

* changed default for send_notifications test to False

* Added two send notifications options. All and some.

* added comma.

* made default arguments for send_notifications.

* default values for notifications set.

* send_emails first version working.

* Fixed reuested changes.

* made minor bug fixes.

* simplify notifications logic, make the email script more systematic

percyliang committed:
set notifications to 2 by default, fix unicode problem

kashizui committed:
Use chown instead of sudo to ensure file permissions

kashizui committed:
Inspect docker logs for hidden failures

kashizui committed:
Documentation

percyliang commented on worker/docker_client.py:
```diff
@@ -174,23 +174,24 @@ def start_container(self, bundle_path, uuid, command, docker_image,
+            # appears in single quotes since we do not. The expansion there,
+            # if any, should happen when bash executes it. Note, since the
+            # user's command can have single quotes we need to escape them.
+            'bash -c '
```
Not having to do `sudo` is actually nicer since some docker images assume that you're running as root, so this is a good change.

percyliang commented on worker/docker_client.py:
```diff
@@ -174,23 +174,24 @@ def start_container(self, bundle_path, uuid, command, docker_image,
+            + '\'(%s) >stdout 2>stderr\'' % command.replace('\'', '\'"\'"\''),
+            # Ensure that any created files are owned by the user/group that
+            # owns the bundle directory, not root.
+            'chown -R $U_ID:$G_ID %s' % docker_bundle_path,
```
Will this command always get executed?  (I think so.)

Has this been tested on the NLP machines with the shared filesystem?  I wonder if we'll always have permissions to do this `chown`...  If this fails, can we make the error go to stderr?  Actually same for the `. $BASHRC` command above...

kashizui replied:
You were right, the chown fails on the NLP machines.
https://codalab.stanford.edu/worksheets/0x69c74999b0e8413eb93d1b0f5ddb6b3d/

As you can see in that bundle, the error gets collected into the `failure_message` field, and same for any resulting error from `. $BASHRC`.

kashizui replied:
Are you saying that if the `chown` fails, we should just pipe the error into `/dev/null`?

percyliang replied:
Definitely not!  We need to be able to change the permissions on the NLP machines to the desired user/group or else the CodaLab user outside docker (not running as root) won't be able access/delete those bundles.

I guess it's a bit strange that 'root' doesn't actually have a superset of the privileges of the user running docker.  Not sure what the best solution here is...

kashizui replied:
I had an earlier solution that created the Docker container to run as the proper user/group, but then running `ldconfig` didn't work because it required root permissions. And `su` complained that it has to be run "in a terminal." Still trying to untangle this..

percyliang reviewed (approved):
LGTM but just want to really make sure that things work before we push this

kashizui committed:
Remove artifacts from stream bytes

kashizui committed:
Experiment: no chown

fabeschan committed:
Backport fix/535 to develop (Worker now prints message on reconnect) (#644)

* remove some codalab/python packages

* add nltk

* Worker now prints message on reconnect (#535)

A worker should print a message when it successfully reconnects and
checks in with the API server after exceptional conditions.

Current behavior:
Worker prints BundleServiceException: Unable to check in with bundle
service repeatedly until reconnected.

Proposed behavior:
Worker additionally prints something like Connected! upon successful
check-in

Fixes #535

percyliang commented:
Another option is to see if we can run CodaLab on the NLP machines with
more root-like permission - probably have to ping sysadmins about that.

On Sat, Jan 21, 2017 at 12:09 PM, Stephen Koo <notifications@github.com>
wrote:

> *@kashizui* commented on this pull request.
> ------------------------------
>
> In worker/docker_client.py
> <https://github.com/codalab/codalab-cli/pull/643>:
>
> > -            # command can have single quotes we need to escape them.
> -            '"[ -e $BASHRC ] && . $BASHRC; "' +
> -            '"cd %s; "' % docker_bundle_path +
> -            '"export HOME=%s; "' % docker_bundle_path +
> -            '\'(%s) >stdout 2>stderr\'' % command.replace('\'', '\'"\'"\''),
> +            # appears in single quotes since we do not. The expansion there,
> +            # if any, should happen when bash executes it. Note, since the
> +            # user's command can have single quotes we need to escape them.
> +            'bash -c '
> +            + '"[ -e $BASHRC ] && . $BASHRC; "'
> +            + '"cd %s; "' % docker_bundle_path
> +            + '"export HOME=%s; "' % docker_bundle_path
> +            + '\'(%s) >stdout 2>stderr\'' % command.replace('\'', '\'"\'"\''),
> +            # Ensure that any created files are owned by the user/group that
> +            # owns the bundle directory, not root.
> +            'chown -R $U_ID:$G_ID %s' % docker_bundle_path,
>
> I had an earlier solution that created the Docker container to run as the
> proper user/group, but then running ldconfig didn't work because it
> required root permissions. And su complained that it has to be run "in a
> terminal." Still trying to untangle this..
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/codalab/codalab-cli/pull/643>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAakuPi0m67IN5aTI3Lo2saFWp8NRER0ks5rUmXrgaJpZM4Lp7ZJ>
> .
>

kashizui committed:
try setting container user

kashizui commented:
I'm not sure if that will necessarily help, since the problem isn't that Docker itself lacks permissions to modify the bundle files.
One solution I'm thinking of is just not to run `ldconfig` inside the container -- if we remove the GPU hacks and only rely on `nvidia-docker`, then we don't need to run `ldconfig`, which requires root.
Then we can directly set the user/group of the process inside the container when we create it using the Docker API without any hacks.

percyliang commented:
Sounds good to me.  How are our GPU hacks (for linking libraries into
docker) different from `nvidia-docker`'s though?

On Sat, Jan 21, 2017 at 1:17 PM, Stephen Koo <notifications@github.com>
wrote:

> I'm not sure if that will necessarily help.
> One solution I'm thinking of is just not to run ldconfig inside the
> container -- if we remove the GPU hacks and only rely on nvidia-docker,
> then we don't need to run ldconfig, which requires root.
> Then we can directly set the user/group of the process inside the
> container when we create it using the Docker API without any hacks.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/codalab/codalab-cli/pull/643#issuecomment-274288891>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAakuKxnxAT3HnkYNBtBAigb-f57KrmFks5rUnXvgaJpZM4Lp7ZJ>
> .
>

kashizui commented:
Main difference is that the hack only mounts a tiny subset of the nvidia libraries (libcuda only), while nvidia-docker does it for all nvidia libraries available on the host, builds a Docker volume for the libraries, creating all the necessary hard links, and even excluding libraries known to cause problems.

percyliang commented:
Ah, cool, thanks.  Btw, did we get nvidia-docker installed on all the jags?

On Sat, Jan 21, 2017 at 1:33 PM, Stephen Koo <notifications@github.com>
wrote:

> Main difference is that the hack only mounts a tiny subset of the nvidia
> libraries (libcuda only), while nvidia-docker does it for all nvidia
> libraries available on the host, builds a Docker volume for the libraries,
> creating all the necessary hard links, and even excluding libraries known
> to cause problems.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/codalab/codalab-cli/pull/643#issuecomment-274289728>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAakuCYVImRuxyChYAvS8FGy8cgDfRwnks5rUnmYgaJpZM4Lp7ZJ>
> .
>

kashizui commented:
Yep with a couple exceptions, I forgot to forward you the email.

kashizui committed:
Remove libcuda hacks

kashizui commented:
Reopening as #647 to merge into develop instead (which has the nvidia-docker support).
